### ISSUE TITLE
FYI: OpenSSL SECLEVEL

### ISSUE THREAD

**Issue by dcooper16** [drwetter/testssl.sh#1433]
OpenSSL 1.1.0 introduced the ability to specify a security level: https://github.com/openssl/openssl/blob/master/doc/man3/SSL_CTX_set_security_level.pod. By default the security level is set to 1 unless a compile-time option is used to set the default a different value. The security level may also be set by the command line, e.g.,
```
openssl s_client -cipher @SECLEVEL=0:ALL -connect 127.0.0.1:443
```
At the moment, testssl.sh does not use the `@SECLEVEL=n` directive, but we may want to look into using `s_client_options()` to add it in some circumstances. The particular reason for suggesting this is https://github.com/openssl/openssl/pull/10786 and https://github.com/openssl/openssl/pull/10787. https://github.com/openssl/openssl/pull/10787 notes that:

> This has as effect that SHA1 and MD5+SHA1 are no longer supported at security level 1. This has as effect that TLS < 1.2 is no longer supported at the default security level of 1, and that you need to set the security level to 0 to use TLS < 1.2.

However, even without https://github.com/openssl/openssl/pull/10786 and https://github.com/openssl/openssl/pull/10787, OpenSSL 1.1.0 and 1.1.1 will test for more weak ciphers if the security level is set to 0 rather than 1.

**Comment by drwetter** [comment]
Thanks, David.

In the script I provided that is already included (https://github.com/drwetter/testssl.sh/blob/3.0/utils/make-openssl111.sh#L17). So as long as people use, we'll be fine.

However that's seldom the case :-)  If I look at the requests here, people use the OpenSSL from the distribution instead. IN terms of security distributors however don't use the flag. So you have a point.

I wasn't aware that one can use it for runtime.

Basically we should still be fine, right?, as most of what's missing then is done in socket automatically -- except in cases where sockets is the only choice.

**Comment by dcooper16** [comment]
> Basically we should still be fine, right?, as most of what's missing then is done in socket automatically

Hi Dirk,

I also have a version of OpenSSL 1.1.1 that I compiled myself, and I compiled it with the default security level set to 0. However, it is more convenient to just use the binary that came with the distribution. :-)

As you noted, in most cases testssl.sh will still produce the correct results even if OpenSSL misses the weak ciphers/protocols, since it will use tls_sockets() to find them. However, for functions like run_allciphers(), testssl.sh will run faster if more of the supported ciphers are found using OpenSSL.

Of course, for the few tests that still only use OpenSSL, if OpenSSL cannot connect, then the results will be incorrect.

I've already developed a PR that changes testssl.sh to use the @SECLEVEL directive when it is supported by OpenSSL. It is a relatively small PR, since changes mainly just needed to be made to s_client_options(), actually_supported_osslciphers(), and listciphers(). It seems to work okay, but since the PR modifies every call to `$OPENSSL s_client` and `$OPENSSL ciphers`, I want to test it very thoroughly before submitting it. So, it is most likely something that won't be merged until after 3.0.

**Comment by Unit193** [comment]
Howdy,

As one of the Debian maintainers of testssl.sh, I'm very interested in this as we have to use the system's OpenSSL binary.  How did the testing on this go?

Thanks!
~Unit 193
Unit193 @ freenode
Unit193 @ OFTC

**Comment by dcooper16** [comment]
Hi @unit193,

I just submitted a PR with the change. I probably should have done it sooner, but I got distracted with other things.

I just did a test with the OpenSSL that comes with my distribution, and it seems that adding "@SECLEVEL=0" may provide very limited benefit. However, OpenSSL does not allow a cipherlist to be just "@SECLEVEL=0". So, whenever a call to `$OPENSSL s_client` would not specify any ciphers, the PR adds `-cipher @SECLEVEL=0:ALL:COMPLEMENTOFALL" to the command line, if "@SECLEVEL=0" is supported. Adding "ALL:COMPLEMENTOFALL" seems to make more of a difference.

Using my distribution's OpenSSL, I get the following results:
```
> openssl ciphers -V -s | wc
     30     240    3034
> openssl ciphers -V -s ALL:COMPLEMENTOFALL | wc
     77     616    7799
> openssl ciphers -V -s @SECLEVEL=0:ALL:COMPLEMENTOFALL | wc
     96     768    9649
```
The only difference between the cipherlist offered by "ALL:COMPLEMENTOFALL" and "@SECLEVEL=0:ALL:COMPLEMENTOFALL" is that the latter includes several ciphers that use either Null authentication or Null encryption.

**Comment by drwetter** [comment]
Thanks, David. Read the PR first. So maybe that's a waste of your energy.

> Using my distribution's OpenSSL, I get the following results

[..]

Same here!

With a manually compiled version 1.1.1 (pull from yesterday):

```
> apps/openssl ciphers -V -s | wc 
     30     240    3034
> apps/openssl ciphers -V -s ALL:COMPLEMENTOFALL | wc   
    110     880   10982
> apps/openssl ciphers -V -s @SECLEVEL=0:ALL:COMPLEMENTOFALL | wc
   110     880   10982
```



```

**Comment by Unit193** [comment]
Howdy,

Thanks very much for the quick turnaround on this!  Fantastic news and can't 
wait for the next release (with this and the license clarification.)

~Unit 193
Unit193 @ freenode
Unit193 @ OFTC
